src/nbl/asset/utils/CSPIRVIntrospector.cpp

```diff
@@ -504,210 +560,333 @@ core::smart_refctd_ptr<const CSPIRVIntrospector::CIntrospectionData> CSPIRVIntro
-        const spirv_cross::SPIRType& mtype = _comp.get_type(_allMembersTypes[m]);
-
-        member.name = _comp.get_member_name(_parentType.self, m);
-        member.size = _comp.get_declared_struct_member_size(_parentType, m);
-        member.offset = _baseOffset + _comp.type_struct_member_offset(_parentType, m);
-        member.rowMajor = _comp.get_member_decoration(_parentType.self, m, spv::DecorationRowMajor);
-        member.type = spvcrossType2E_TYPE(mtype.basetype);
-        member.arrayStride = 0u;
-
-        // if array, then we can get array stride from decoration (via spirv-cross)
-        // otherwise arrayStride is left with value 0
-        if (mtype.array.size())
+    auto addBaseAndConvertTypeToImmutable = [&](SType<true>* type)->void
+    {
+        auto* outImmutable = reinterpret_cast<SType<false>*>(type);
+        addBaseAndConvertStringToImmutable(outImmutable->typeName);
+        addBaseAndConvertExtentToImmutable(outImmutable->count);
+        // get these before they change
+        const auto memberTypes = type->memberTypes()(basePtr);
+        const auto memberNames = type->memberNames()(basePtr);
+        for (auto m=0; m<type->memberCount; m++)
         {
-            member.count = mtype.array[0];
-            member.arrayStride = _comp.type_struct_member_array_stride(_parentType, m);
-            member.countIsSpecConstant = !mtype.array_size_literal[0];
-            if (member.countIsSpecConstant)
-			{
-                auto sc_itr = _mapId2sconst.find(member.count);
-                assert(sc_itr!=_mapId2sconst.end());
-                auto sc = sc_itr->second;
-                member.count_specID = sc->id;
-            }
+            reinterpret_cast<const SType<false>**>(memberTypes)[m] = reinterpret_cast<const SType<false>*>(memberTypes[m](basePtr));
+            reinterpret_cast<std::span<const char>*>(memberNames)[m] = memberNames[m](basePtr);
         }
+        outImmutable->memberInfoStorage = type->memberInfoStorage(basePtr);
+    };
+    auto addBaseAndConvertBlockToImmutable = [&](SMemoryBlock<true>& block)->void
+    {
+        visitMemoryBlockPreOrderDFS(block,addBaseAndConvertTypeToImmutable);
+        auto& asImmutable = reinterpret_cast<SMemoryBlock<false>&>(block);
+        asImmutable.type = reinterpret_cast<SType<false>*>(block.type(basePtr));
+    };
 
-        if (mtype.basetype == spirv_cross::SPIRType::Struct) //recursive introspection done in DFS manner (and without recursive calls)
-            _pushStack.push({member.members, mtype, member.offset});
-        else
-		{
-            member.mtxRowCnt = mtype.vecsize;
-            member.mtxColCnt = mtype.columns;
-            if (member.mtxColCnt > 1u)
-                member.mtxStride = _comp.type_struct_member_matrix_stride(_parentType, m);
+    addBaseAndConvertBlockToImmutable(reinterpret_cast<SPushConstantInfo<true>&>(m_pushConstants));
+    addBaseAndConvertStringToImmutable(m_pushConstants.name);
+    std::ostringstream debug = {};
```

Comment:
time to move all the debug print to a dedicated method that takes a `system::ILogger*`